### PR TITLE
Add L postfix for hex string

### PR DESCRIPTION
**Pull request by dkorpel**
Since we fixed the 'w' and 'd' postfixes for hex strings (https://github.com/dlang/dmd/pull/16118), it would make sense to add a postfix for size 8 as well. It could replace the `cast(immutable ulong[])` syntax, which can be confusing, since it differs between CTFE (big-endian) and run-time (native-endian).

@WalterBright said "Sounds good" to this proposal, though perhaps community feedback is also in place.

@tgehr In the last monthly meeting you suggested not using cast syntax.

### CODE REVIEW

File: changelog/dmd.hexstring-cast.dd [hunk 1]
```diff
@@ -11,13 +11,17 @@ immutable uint[] data = cast(immutable uint[]) x"AABBCCDD";
...
-Character postfixes can now also be used for integers of size 2 or 4:
+Character postfixes can now also be used for integers of size 2 or 4, while an `L` postfix can be used for size 8:
 ---
 immutable ushort[] f = x"80 3F"w;
 static assert(f[0] == 0x803F);
 
-immutable int[] f = x"80 35 FF FD"d;
-static assert(f[0] == 0x803FFF);
+immutable int[] g = x"80 35 FF FD"d;
+static assert(g[0] == 0x803FFF);
```
Comment (ibuclaw) [inline changelog/dmd.hexstring-cast.dd:20]:
```suggestion
immutable int[] g = x"80 35 FF FD"d;
static assert(g[0] == 0x8035FFFD);
```
This static assert doesn't match the assigned value.

File: compiler/test/runnable/literal.d [hunk 2]
```diff
@@ -243,23 +243,26 @@ void test12950()
...
     static assert(x[0] == 0xFFAADDEE);
     assert(x[0] == 0xFFAADDEE);
 
-    static immutable ulong[] y = cast(immutable ulong[]) x"1122334455667788AABBCCDDEEFF0099";
+    static assert(is(typeof(x""L) == immutable(ulong)[]));
+    static immutable ulong[] y = x"1122334455667788AABBCCDDEEFF0099"L;
     static assert(y[0] == 0x1122334455667788);
     static assert(y[1] == 0xAABBCCDDEEFF0099);
     assert(y[0] == 0x1122334455667788);
     assert(y[1] == 0xAABBCCDDEEFF0099);
```
Comment (ibuclaw) [inline compiler/test/runnable/literal.d:255]: What about assigning hexstrings to local variables?

**Review by ibuclaw — DISMISSED** [review]

File: compiler/test/runnable/literal.d (hunk 2 above)
Comment (dkorpel) [reply compiler/test/runnable/literal.d:255]: Added a test

**Review by thewilsonator — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

File: changelog/dmd.hexstring-cast.dd [hunk 3]
```diff
@@ -11,13 +11,17 @@ immutable uint[] data = cast(immutable uint[]) x"AABBCCDD";
...
-Character postfixes can now also be used for integers of size 2 or 4:
+Character postfixes can now also be used for integers of size 2 or 4, while an `L` postfix can be used for size 8:
 ---
 immutable ushort[] f = x"80 3F"w;
 static assert(f[0] == 0x803F);
 
-immutable int[] f = x"80 35 FF FD"d;
-static assert(f[0] == 0x803FFF);
+immutable int[] g = x"80 35 FF FD"d;
+static assert(g[0] == 0x803FFFFD);
```
Comment (ibuclaw) [inline changelog/dmd.hexstring-cast.dd:20]:
```suggestion
immutable int[] g = x"80 35 FF FD"d;
static assert(g[0] == 0x8035FFFD);
```

**Comment by PetarKirov** [comment]
Using the char postfixes for binary data and the `L` which so far meant only `long` may be confusing to many.
How about using a suffix like `u8`, `u16`, `u32`, `u64`, `u128`, `u256`, etc.? If DMD were to support C23 in ImportC, it will need to support bit-precise integer types, see https://blog.tal.bi/posts/c23-bitint/

**Comment by dkorpel** [comment]
I agree that the char postfixes aren't ideal, and if I were to redesign hex strings from scratch, I wouldn't give hex strings a string type in the first place. However, my goal is to speed up compilation time of large `ulong` tables with minimal language additions, and the string postfixes is what we have currently.

I didn't want to add an 'L' postfix either, but I'm not a fan of the `cast(immutable ulong[])` syntax either because cast is a blunt tool, plus the aforementioned endian inconsistency. 

However, I just now realized we actually don't need an L suffix! Consider how regular strings rarely need a postfix:

```D
wstring ws = "abc"w;
wstring ws1 = "abc";
dstring ds = "abc";

wstring ws2 = "abc"c; // Error: char postfix commits to string instead of wstring
```

Hex strings could work the same way.
```D
wstring ss = x"FFFF"w;
immutable ushort[] ss = x"FFFF";
immutable long[] = x"FFFF FFFF FFFF FFFF";

wstring ss = x"FFFF"c; // Error: char postfix commits to size 1
```

No grammar update needed, and it feels really obvious.

**Comment by PetarKirov** [comment]
> Consider how regular strings rarely need a postfix [..]

This is how arrays in general work, so obviously 👍 from me.

```d
ubyte[] a = [1,2,3,4,5];
ushort[] b = [1,2,3,4,5];
uint[] c = [1,2,3,4,5];
ulong[] d = [1,2,3,4,5];
float[] e = [1,2,3,4,5];
double[] f = [1,2,3,4,5];
real[] g = [1,2,3,4,5];
```

**Pull request closed without merging**